### PR TITLE
[Diffsuion & JIT_kernel] QKNorm cross heads kernel

### PR DESCRIPTION
**Pull request by BBuf**
## Motivation

Refer to the implementation at [here](https://github.com/sgl-project/sglang/blob/main/python/sglang/jit_kernel/csrc/elementwise/fused_add_rmsnorm.cuh) , which implements a cross-head QK Norm kernel commonly used in diffusion models such as [Wan](https://github.com/sgl-project/sglang/blob/main/python/sglang/multimodal_gen/runtime/models/dits/wanvideo.py#L336-L338), LTX2(https://github.com/sgl-project/sglang/blob/main/python/sglang/multimodal_gen/runtime/models/dits/ltx_2.py#L489-L502) video model. 

Currently, SGLang Diffusion employs a standard PyTorch implementation. Based on the micro-benchmark below, applying this optimization should yield benefits for Diffusion model applications. 

### H200

```shell
 sglang python3 /home/sglang/bbuf/sglang/python/sglang/jit_kernel/benchmark/bench_qknorm_across_heads.py
qknorm-across-heads-performance:
    batch_size  hidden_dim  SGL JIT Kernel  SGL AOT Kernel  FlashInfer     PyTorch
0          1.0       512.0        1.408159        1.934164    2.878205    2.396000
1          1.0      1024.0        1.604982        1.943352    2.856166    3.120050
2          1.0      2048.0        1.738031        1.970768    2.937277    4.003000
3          1.0      4096.0        2.044754        2.150978    3.151675    5.668969
4          1.0      8192.0        2.687469        2.542889    3.625002    8.812829
5          2.0       512.0        1.602289        1.947592    2.890425    3.048249
6          2.0      1024.0        1.607094        1.926616    2.867172    3.878067
7          2.0      2048.0        1.742574        1.979103    2.953176    5.779670
8          2.0      4096.0        2.062239        2.131626    3.165222    9.287712
9          2.0      8192.0        2.700331        2.550890    3.683352   16.428769
10         4.0       512.0        1.577747        1.949676    2.895959    2.997315
11         4.0      1024.0        1.604107        1.936889    2.877497    3.967130
12         4.0      2048.0        1.754672        1.992359    2.900893    5.760394
13         4.0      4096.0        2.077435        2.155970    3.157134    9.330296
14         4.0      8192.0        2.731756        2.622789    3.671257   16.394464
15         8.0       512.0        1.609985        1.990196    2.941532    3.125523
16         8.0      1024.0        1.630004        1.995473    2.918256    4.038142
17         8.0      2048.0        1.801039        2.060982    2.977578    5.852311
18         8.0      4096.0        2.167062        2.238259    3.239715    9.367947
19         8.0      8192.0        2.829702        2.629480    3.767445   16.461402
20        16.0       512.0        1.636716        2.061677    2.988121    3.171746
21        16.0      1024.0        1.658844        2.050617    2.968823    4.055781
22        16.0      2048.0        1.813107        2.103305    3.046661    5.866874
23        16.0      4096.0        2.153550        2.271488    3.312413    9.364856
24        16.0      8192.0        2.865404        2.669108    3.790158   16.536922
25        32.0       512.0        1.646796        2.072678    3.015226    3.212402
26        32.0      1024.0        1.682800        2.099699    2.991142    4.135111
27        32.0      2048.0        1.837501        2.138137    3.090002    5.997647
28        32.0      4096.0        2.209491        2.302114    3.339683    9.564086
29        32.0      8192.0        2.958240        2.744394    3.921563   16.835010
30        64.0       512.0        1.697991        2.106442    3.057419    3.256840
31        64.0      1024.0        1.718147        2.110734    3.053078    4.155773
32        64.0      2048.0        1.891098        2.175374    3.150112    5.966082
33        64.0      4096.0        2.261412        2.383939    3.423071    9.611928
34        64.0      8192.0        3.123701        3.019044    4.024222   17.015155
35       128.0       512.0        1.784741        2.190324    3.236028    3.454331
36       128.0      1024.0        1.899194        2.235239    3.225206    4.461556
37       128.0      2048.0        2.122253        2.359497    3.304758    6.200198
38       128.0      4096.0        2.691244        2.871827    3.634920    9.885928
39       128.0      8192.0        3.598804        3.837579    4.563051   17.136435
40       256.0       512.0        1.844194        2.347478    3.448340    3.602453
41       256.0      1024.0        2.073799        2.375687    3.512117    4.574109
42       256.0      2048.0        2.500434        2.686402    3.635724    6.240418
43       256.0      4096.0        3.307879        3.440966    4.210651   11.136000
44       256.0      8192.0        6.437042        5.788892    5.893284   20.294239
45       512.0       512.0        1.936333        2.821110    3.793886    3.529261
46       512.0      1024.0        2.377824        2.882018    3.911659    4.632192
47       512.0      2048.0        3.236985        3.533543    4.401111    6.839170
48       512.0      4096.0        6.071986        5.276495    5.832248   12.316856
49       512.0      8192.0       11.541461       11.398904   12.458365   20.401542
50      1024.0       512.0        2.461973        4.141131    5.081170    4.473544
51      1024.0      1024.0        3.168057        4.598080    5.607329    6.096386
52      1024.0      2048.0        4.960575        5.319408    6.109368    9.790194
53      1024.0      4096.0        9.341096        9.662918   10.176309   16.714387
54      1024.0      8192.0       19.115782       20.043544   20.267191   31.437784
55      2048.0       512.0        3.167390        7.007722    8.212610    7.008000
56      2048.0      1024.0        4.950328        6.698104    9.094531    9.709676
57      2048.0      2048.0        7.739965        8.964276    9.746161   15.758207
58      2048.0      4096.0       16.552114       17.585487   18.641672   28.146625
59      2048.0      8192.0       50.998329       51.091578   53.892641   70.593621
60      4096.0       512.0        4.563406       11.637444   12.899523   11.188669
61      4096.0      1024.0        7.440724       11.859483   12.947170   17.083533
62      4096.0      2048.0       13.772873       16.862922   17.865829   28.968231
63      4096.0      4096.0       42.739338       46.910312   48.848851   65.082792
64      4096.0      8192.0       97.908767      100.183472  101.216002  126.576759
65      8192.0       512.0        7.107411       21.572598   22.720099   19.963824
66      8192.0      1024.0       12.772678       21.902188   23.118611   31.420746
67      8192.0      2048.0       35.692123       42.479245   44.442240   59.982188
68      8192.0      4096.0       80.484846       90.773844   92.843732  114.565076
69      8192.0      8192.0      192.465100      197.108366  199.801859  232.331049
```

af95f26c-1489-4012-a826-cbc8f4e56ebb

We can use 128 bits vectorize load/store .

### B200

```shell
bbuf python3 /home/lmsys/bbuf/sglang/python/sglang/jit_kernel/benchmark/bench_qknorm_across_heads.py
qknorm-across-heads-performance:
    batch_size  hidden_dim  SGL JIT Kernel  SGL AOT Kernel  FlashInfer     PyTorch
0          1.0       512.0        1.778977        2.453934    3.386504    3.048790
1          1.0      1024.0        1.992154        2.459118    3.356109    3.968000
2          1.0      2048.0        2.095112        2.412638    3.424000    4.924379
3          1.0      4096.0        2.399849        2.616545    3.541271    6.937758
4          1.0      8192.0        3.005666        3.077995    4.150158   10.603095
5          2.0       512.0        2.008275        2.410920    3.422718    3.807407
6          2.0      1024.0        2.011846        2.327062    3.357288    4.563200
7          2.0      2048.0        2.104498        2.434673    3.425297    6.634842
8          2.0      4096.0        2.410935        2.557339    3.554275   10.469671
9          2.0      8192.0        3.024054        3.006745    4.191846   18.029935
10         4.0       512.0        1.995928        2.376279    3.413257    3.605726
11         4.0      1024.0        2.016894        2.374227    3.347013    4.730243
12         4.0      2048.0        2.091917        2.409728    3.405568    6.607668
13         4.0      4096.0        2.435349        2.551068    3.553200   10.417149
14         4.0      8192.0        3.054325        2.991962    4.182475   18.016865
15         8.0       512.0        2.002970        2.394351    3.422450    3.778595
16         8.0      1024.0        2.021151        2.351469    3.381410    4.731775
17         8.0      2048.0        2.108950        2.406275    3.424470    6.609455
18         8.0      4096.0        2.436468        2.540635    3.554879   10.405219
19         8.0      8192.0        3.045227        2.986400    4.191406   18.018404
20        16.0       512.0        2.010722        2.389976    3.390350    3.744083
21        16.0      1024.0        2.026849        2.369734    3.374912    4.790141
22        16.0      2048.0        2.106578        2.415115    3.409307    6.607913
23        16.0      4096.0        2.425067        2.585421    3.569938   10.407130
24        16.0      8192.0        3.028023        3.028547    4.199525   18.013391
25        32.0       512.0        2.003207        2.406547    3.429579    3.789990
26        32.0      1024.0        2.022806        2.367692    3.376740    4.772322
27        32.0      2048.0        2.107491        2.430066    3.435639    6.659527
28        32.0      4096.0        2.440607        2.577812    3.583466   10.464770
29        32.0      8192.0        3.059654        3.018160    4.206111   18.081559
30        64.0       512.0        2.019542        2.435109    3.410054    3.814720
31        64.0      1024.0        2.024834        2.402146    3.399255    4.798514
32        64.0      2048.0        2.131743        2.465489    3.476920    6.675167
33        64.0      4096.0        2.538236        2.624000    3.619006   10.505989
34        64.0      8192.0        3.137608        3.069784    4.269585   18.104128
35       128.0       512.0        2.035314        2.543843    3.485600    3.903262
36       128.0      1024.0        2.066437        2.520925    3.458377    4.875157
37       128.0      2048.0        2.203311        2.633356    3.517615    6.775467
38       128.0      4096.0        2.562768        2.870644    3.698571   10.577012
39       128.0      8192.0        3.196617        3.880485    4.357266   18.264385
40       256.0       512.0        2.092022        2.621568    3.636092    4.068571
41       256.0      1024.0        2.157444        2.637281    3.654196    5.120626
42       256.0      2048.0        2.425076        2.826978    3.736095    7.212022
43       256.0      4096.0        3.019628        3.818440    4.183633   11.290606
44       256.0      8192.0        4.410286        6.048971    6.075422   19.571097
45       512.0       512.0        2.200174        3.016665    4.172208    4.369720
46       512.0      1024.0        2.426360        3.048473    4.240378    5.516838
47       512.0      2048.0        2.932680        3.570607    4.436830    7.702435
48       512.0      4096.0        4.077151        5.394187    6.188110   12.175064
49       512.0      8192.0        7.359695       10.436571   10.819934   21.040084
50      1024.0       512.0        2.403922        4.155100    5.337947    5.079872
51      1024.0      1024.0        2.773954        4.045726    5.526315    6.638568
52      1024.0      2048.0        3.619861        5.257059    5.671835    9.718948
53      1024.0      4096.0        6.200674        9.159154    9.291877   15.905626
54      1024.0      8192.0       12.561960       18.740202   18.667693   29.382651
55      2048.0       512.0        2.819441        6.136898    7.640274    7.110091
56      2048.0      1024.0        3.592000        6.157080    7.821968    9.902161
57      2048.0      2048.0        5.770609        8.833867    9.219396   15.569224
58      2048.0      4096.0       10.226307       16.509656   16.872504   26.829058
59      2048.0      8192.0       22.705137       35.153707   35.257032   50.203771
60      4096.0       512.0        4.087836       10.319042   11.976608   11.300170
61      4096.0      1024.0        5.656288       10.774708   12.062489   16.920000
62      4096.0      2048.0        9.432856       15.765213   16.117311   28.104440
63      4096.0      4096.0       18.512054       31.320659   31.464013   49.698361
64      4096.0      8192.0       47.742401       81.381871   82.375080  108.461098
65      8192.0       512.0        6.313091       18.827515   20.372990   19.426150
66      8192.0      1024.0        9.255502       19.305035   20.478439   30.176093
67      8192.0      2048.0       16.870470       29.852832   30.115987   51.809063
68      8192.0      4096.0       38.851415       69.922861   71.422018  101.826376
69      8192.0      8192.0       91.878096      160.443314  158.155705  198.531171
```

dbeda512-a744-45c4-ac93-14597411b9f5

We can use 256-bits vectorize load/store in B200.

## Limitation

Currently, compared to the single rmsnorm kernel, the occupancy is not high because more registers are used (increased from 40 to 48). Maybe, there should still be room for optimization. 

```python
import itertools
from typing import Tuple

import torch
import triton
import triton.testing
from sgl_kernel import rmsnorm

from sglang.jit_kernel.benchmark.utils import is_in_ci
from sglang.jit_kernel.norm import fused_inplace_qknorm_across_heads
from sglang.srt.utils import get_current_device_stream_fast

IS_CI = is_in_ci()

alt_stream = torch.cuda.Stream()


def sglang_jit_qknorm_across_heads(
    q: torch.Tensor,
    k: torch.Tensor,
    q_weight: torch.Tensor,
    k_weight: torch.Tensor,
) -> None:

    fused_inplace_qknorm_across_heads(q, k, q_weight, k_weight)


DTYPE = torch.bfloat16
DEVICE = "cuda"

batch_size = 64
hidden_size = [512, 1024]

for hidden_dim in hidden_size:
    q = torch.randn((batch_size, hidden_dim), dtype=DTYPE, device=DEVICE)
    k = torch.randn((batch_size, hidden_dim), dtype=DTYPE, device=DEVICE)
    q_weight = torch.randn(hidden_dim, dtype=DTYPE, device=DEVICE)
    k_weight = torch.randn(hidden_dim, dtype=DTYPE, device=DEVICE)
    
    sglang_jit_qknorm_across_heads(q, k, q_weight, k_weight)
    print('passed')

```

34e6a974-3089-4e83-9ffe-5485284d4876

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/jit_kernel/csrc/elementwise/qknorm_across_heads.cuh [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+  }
+
+  auto cg_warp = cooperative_groups::tiled_partition<32>(cooperative_groups::this_thread_block());
+  float* buffer_q = shared_memory;       // [0, 31] for Q
+  float* buffer_k = shared_memory + 32;  // [32, 63] for K
+
+  // ========== Reduction phase: Compute rsqrt for both Q and K ==========
+
+  // Step 0: Warp Reduce for Q
+  float warp_sum_q =
```
Comment (HydraQYH) [inline python/sglang/jit_kernel/csrc/elementwise/qknorm_across_heads.cuh:113]: For diffusion models, what is the typical range of its `hidden_dim`? If `hidden_dim` is small enough that a single warp can handle all elements, then we only need one warp reduce here. For example, if `hidden_dim` is 512, and the kernel is running on Blackwell.

**Review by HydraQYH — COMMENTED** [review]
Just one question here.

File: python/sglang/jit_kernel/csrc/elementwise/qknorm_across_heads.cuh (hunk 1 above)
Comment (BBuf) [reply python/sglang/jit_kernel/csrc/elementwise/qknorm_across_heads.cuh:113]: 1536, 3072, 5120 is the most common shape. refer to https://acntwz63wyi6.feishu.cn/wiki/XRfdwY9jFii2MKkTMGwcRJBhnuf

Comment (HydraQYH) [reply python/sglang/jit_kernel/csrc/elementwise/qknorm_across_heads.cuh:113]: Got it.

**Review by HydraQYH — APPROVED** [review]
